homeassistant/components/bosch_shc/config_flow.py

```diff
@@ -0,0 +1,220 @@
+    session = await hass.async_add_executor_job(
+        SHCSession,
+        host,
+        cert,
+        key,
+        True,
+        zeroconf,
+    )
+
+    await hass.async_add_executor_job(session.authenticate)
+
+
+def write_tls_asset(hass: core.HomeAssistant, filename: str, asset: bytes):
+    """Write the tls assets to disk."""
+    makedirs(hass.config.path(DOMAIN), exist_ok=True)
+    with open(hass.config.path(DOMAIN, filename), "w") as file_handle:
+        file_handle.write(asset.decode("utf-8"))
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Bosch SHC."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+    info = None
+    host = None
+    hostname = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            try:
+                self.info = info = await self._get_info(host)
+            except SHCConnectionError:
+                errors["base"] = "cannot_connect"
+            except SHCmDNSError:
+                _LOGGER.warning("Error looking up mDNS entry")
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(info["unique_id"])
+                self._abort_if_unique_id_configured({CONF_HOST: host})
+                self.host = host
+                return await self.async_step_credentials()
+
+        return self.async_show_form(
+            step_id="user", data_schema=HOST_SCHEMA, errors=errors
+        )
+
+    async def async_step_credentials(self, user_input=None):
+        """Handle the credentials step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                helper = await self.hass.async_add_executor_job(
+                    SHCRegisterClient, self.host, user_input[CONF_PASSWORD]
+                )
+                result = await self.hass.async_add_executor_job(
+                    helper.register, self.host, "HomeAssistant"
+                )
+                if "token" in result:
+                    self.hostname = result["token"].split(":", 1)[1]
+
+                await self.hass.async_add_executor_job(
+                    write_tls_asset, self.hass, CONF_SHC_CERT, result["cert"]
+                )
+                await self.hass.async_add_executor_job(
+                    write_tls_asset, self.hass, CONF_SHC_KEY, result["key"]
+                )
+                await validate_input(
+                    self.hass,
+                    self.host,
+                    self.hass.config.path(DOMAIN, CONF_SHC_CERT),
+                    self.hass.config.path(DOMAIN, CONF_SHC_KEY),
+                )
```

Comment:
We should merge all six executor calls here into one call. Extract the sync calls into a function that we schedule once on the executor.